src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
 
             stateMachineType = new IteratorStateMachine(slotAllocatorOpt, compilationState, method, methodOrdinal, isEnumerable, elementType);
             compilationState.ModuleBuilderOpt.CompilationState.SetStateMachineType(method, stateMachineType);
-            return new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics).Rewrite();
+            var rewriter = new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics);
+            if (!rewriter.VerifyPresenceOfRequiredAPIs())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        // returns true if all types we need are present
+        protected bool VerifyPresenceOfRequiredAPIs()
+        {
+            bool hasErrors = false;
+
+            EnsureSpecialType(SpecialType.System_Int32, ref hasErrors);
+            EnsureSpecialMember(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator
+            EnsureSpecialType(SpecialType.System_Collections_IEnumerator, ref hasErrors);
+            EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__Current, ref hasErrors);
+            EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__MoveNext, ref hasErrors);
+            EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__Reset, ref hasErrors);
+
+            // IEnumerator<T>
+            EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
+            EnsureSpecialMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current, ref hasErrors);
+
+            if (_isEnumerable)
+            {
+                // IEnumerable and IEnumerable<T>
+                EnsureSpecialType(SpecialType.System_Collections_IEnumerable, ref hasErrors);
+                EnsureSpecialMember(SpecialMember.System_Collections_IEnumerable__GetEnumerator, ref hasErrors);
+                EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T, ref hasErrors);
+                EnsureSpecialMember(SpecialMember.System_Collections_Generic_IEnumerable_T__GetEnumerator, ref hasErrors);
+            }
+
+            return !hasErrors;
+        }
+
+        private void EnsureSpecialMember(SpecialMember member, ref bool hasErrors)
+        {
+            Symbol symbol;
+            if (!Binder.TryGetSpecialTypeMember(F.Compilation, member, body.Syntax, diagnostics, out symbol))
+            {
+                hasErrors = true;
+            }
```

Comment:
`hasErrors` is unchanged if the symbol has use-site errors. That's different from the behavior of `EnsureSpecialType` and the VB version of this method. Is that intentional? (Do we want to rewrite in that case?)